### ISSUE TITLE
authenticateAccount fails to login a user

### ISSUE THREAD

**Issue by xtu** [stormpath/express-stormpath#217]
I use ```app.createAccount``` to create an account and then use ```app.authenticateAccount``` to try to login the user. An error returns as follows:

ResourceError: HTTP 400, Stormpath 7104 (http://docs.stormpath.com/errors/7104): Login attempt failed because there is no Account in the Application's associated Account Stores with the specified username or email.

**Comment by robertjd** [comment]
Could you show me the code that you are using for this operations?  I would like to ensure that the parameters to `authenticateAccount` are correct, and that `app` is a proper reference.  Thanks!

**Comment by xtu** [comment]
I am using express-stormpath 2.0.0. Here is the code I've used:
```javascript
var app = req.app.get('stormpathApplication');
app.createAccount(account, function (err, account) {
  if (err) throw err;
  app.authenticateAccount({
      email: req.form.email,
      password: req.form.password,
  }, function (err, result) {
      if (err) {
          log.error('failed to login in the user after registration: ERROR=' + err)
      }
      res.redirect('/')
  });
});
```

**Comment by robertjd** [comment]
Thanks for the code!  The `authenticateAccount` method expects the parameter `username`, so you'll need to change that to look like this:

```
app.authenticateAccount({
    username: req.form.email,
    password: req.form.password,
}, function (err, result) {
```

It's still okay to pass the email as the username, the Stormpath API will accept both for the username field.

Please note that the 2.x series of this module uses a cookie-based token strategy for authentication (if you are using our default `/login` route).  With your solution those cookies are not created, are you intending to implement a different session mechanism?

Thanks,
Robert

**Comment by xtu** [comment]
Hi Robert, after change the parameter name to ```username```, I am still not able to log in automatically. Is it because the cookies are not created? I am not intending to implement a different session mechanism. I just want to automatically log in after registration.

**Comment by robertjd** [comment]
Correct, the cookies won't be created unless you do that manually.  If you want to log the user in after registration, you can use the auto-login option of the registration configuration.  That looks like this:

```
app.use(stormpath.init(app, {
  website: true,
  expand: {
    customData: true
  },
  web: {
    register: {
      autoLogin: true
    }
  }
}));
```

The auto-login documentation is here: http://docs.stormpath.com/nodejs/express/latest/registration.html#auto-login

**Comment by xtu** [comment]
I have tried to enable ```autoLogin```, still does not work. After registration, it navigates to the login page automatically. I used ```stormpathPath.loginRequired``` to protect my pages.

**Comment by robertjd** [comment]
Hi Xtu, at this point it would be helpful if I could see your code and how our middleware is connected.  If you can share here (or on a gist) that would be great, otherwise please contact us via support@stormpath.com and we can share code privately.  Thanks!

**Comment by xtu** [comment]
I have read the source code included in my ```node-modules``` folder for ```stormpath```. I can see that following function does not use ```autoLogin``` option at all. Does this help on investigate the problem. I   am currently working on a separate feature and not able to share the code.

```javascript
Application.prototype.authenticateAccount = function authenticateApplicationAccount(authcRequest, callback) {
  var _this = this,
    username = authcRequest.username,
    password = authcRequest.password,
    type = authcRequest.type || 'basic';
  var accountStore = ('string' === typeof authcRequest.accountStore) ?
    {href: authcRequest.accountStore} :
    authcRequest.accountStore;

  var loginAttempt = {
    type: type,
    value: utils.base64.encode(username + ":" + password)
  };

  if (authcRequest.accountStore){
    loginAttempt.accountStore = accountStore;
  }

  _this.dataStore.createResource(
    _this.loginAttempts.href,
    {expand: 'account'},
    loginAttempt,
    AuthenticationResult,
    function(err,authenticationResult){
      if(err){
        callback(err);
      }
      else{
        authenticationResult.application = _this;
        callback(null,authenticationResult);
      }
  });
};
```

**Comment by robertjd** [comment]
Hi Xtu, that is the code for the underlying Node SDK.  It does not handle the auto-login feature.  The auto login feature is handled here, in this library:  https://github.com/stormpath/express-stormpath/blob/master/lib/controllers/register.js#L130

Can you confirm that the email verification feature is disabled on the Stormpath Directory that you are using with this application?  If email verification is enabled, the auto login feature cannot work.

Here is a fork of our example project, which shows how to use the auto login feature:

https://github.com/robertjd/express-stormpath-sample-project/tree/auto-login-with-registration

Hope this helps,
Robert

**Comment by xtu** [comment]
Hi @robertjd , I am using my own register view and logics to register a user, so I have not been using that ```register.js```. Could you tell me how to create that cookie session manually if it is mandatory to log in a user?

I can confirm that the email verification is disabled.

**Comment by robertjd** [comment]
To create the cookies manually, you will need to invoke the `createSession` method, as you can see here:

https://github.com/stormpath/express-stormpath/blob/fe95e0cb83d75643e6521aeddd4795167c18798e/lib/controllers/register.js#L135

You can require that function like this:

```
var createSession = require('express-stormpath/lib/helpers/create-session.js');
```

It needs to be called with a password grant authentication result, because that result contains the Oauth2 tokens that we use in the cookies.

Have to looked at our post-registration handler?  It may help you out:

http://docs.stormpath.com/nodejs/express/latest/registration.html#post-registration-handler

If that is not useful, could you let us know what would be more helpful?  We're always looking for ways to make the library better.  For example, would a pre-registration handler be better?

Thank you for all the information so far!

**Comment by xtu** [comment]
Can I get that password grant authentication result from any of post-registration handler parameters?

Pre-registration handler looks quite useful to me. I think users may have their own business logics before registration in most cases.
